File: nla/python/steepest-descent.py
import numpy as np

def steepest_descent(A,b,x,tol):
  r = b
  r -= A @ x
  while True:
    r2 = np.dot(r,r)
    if (r2<tol*tol):
      break
    p = A @ r
    alpha = (r2)/(np.dot(p,r))
    x += alpha*r
    r -= alpha*p

File: nla/code/cpp/steepest_descent.cc
void steepest_descent(arma::Col<double>& x,
		      const arma::Mat<double>& A,
		      const arma::Col<double>& b,
		      double tolerance)
{
  arma::Col<double> r(b);
  arma::Col<double> p(size(x));
  p = A*x;
  r -= d;
  double res = arma::norm(r);
  while (res > tolerance) {
    p = A*r;
    double alpha = (res*res)/dot(p,r);
    // Breakdown not caught
    x += alpha*r;
    r -= alpha*p;
  }
}
